engine/resources/pippet_run.go

```diff
@@ -0,0 +1,160 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+)
+
+// PippetRunner is the interface to the wrapped "puppet yamlresource receive"
+// process that keeps running alongside mgmt. Individual resources are expected
+// to register themselves, and can use the Apply function from that point on.
+type PippetRunner interface {
+	Register(interface{}) error
+	Unregister(interface{}) error
+	Apply(string, string, string) PippetResult
```

Comment:
Weird return arg, would: (PippetResult, error) be more appropriate?

Reply:
It's a good point. The result from Puppet's sync operation is unmarshalled from JSON, in PippetResult form, and that *has* to include any errors that happened (we have no other way to get diagnostics from Puppet to mgmt).

Of course, we could go ahead and extract the interesting fields within the Apply function, and then return something that needs not be a PippetResult plus an error. Would you prefer that?

Reply:
I think you want (PippetResult, error) here... Let's see what that looks like.

Reply:
I reworked the function, and it became apparent that returning (bool, error) is sufficient (I feel).

Let me know what you think. Keeping this in a separate commit for now, so you can review the change itself if you wish.